### PR TITLE
Rework gRPC certificate loading one last time

### LINKED ISSUE #2604: Firestore fails to find certificates on macOS
Pod installing Firestore ends up crashing with this message:

```
2019-03-21 10:06:37.143496-0700 macOS_example[39141:28967486] *** Assertion failure in std::string firebase::firestore::remote::LoadGrpcRootCertificate()(), /Users/mcg/src/firebase/ios-sdk/Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder_apple.mm:80
2019-03-21 10:06:37.148545-0700 macOS_example[39141:28967486] [General] An uncaught exception was raised
2019-03-21 10:06:37.148562-0700 macOS_example[39141:28967486] [General] FIRESTORE INTERNAL ASSERTION FAILED: Could not load root certificates from the bundle. SSL cannot work. (expected path)
2019-03-21 10:06:37.148636-0700 macOS_example[39141:28967486] [General] (
	0   CoreFoundation                      0x00007fff4c9f8ded __exceptionPreprocess + 256
	1   libobjc.A.dylib                     0x00007fff78ac4720 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff4ca13a4a +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x00007fff4ee06eed -[NSAssertionHandler handleFailureInFunction:file:lineNumber:description:] + 166
	4   macOS_example                       0x0000000100268675 _ZN8firebase9firestore4util8internal4FailEPKcS4_iRKNSt3__112basic_stringIcNS5_11char_traitsIcEENS5_9allocatorIcEEEE + 533
	5   macOS_example                       0x0000000100268d32 _ZN8firebase9firestore4util8internal4FailEPKcS4_iRKNSt3__112basic_stringIcNS5_11char_traitsIcEENS5_9allocatorIcEEEES4_ + 1538
	6   macOS_example                       0x000000010024d6d6 _ZN8firebase9firestore6remote23LoadGrpcRootCertificateEv + 118
	7   macOS_example                       0x000000010023f320 _ZNK8firebase9firestore6remote14GrpcConnection13CreateChannelEv + 96
	8   macOS_example                       0x000000010023ee3b _ZN8firebase9firestore6remote14GrpcConnection16EnsureActiveStubEv + 267
	9   macOS_example                       0x0000000100241a95 _ZN8firebase9firestore6remote14GrpcConnection12CreateStreamEN4absl11string_viewERKNS0_4auth5TokenEPNS1_18GrpcStreamObserverE + 85
	10  macOS_example                       0x0000000100341a25 _ZN8firebase9firestore6remote11WriteStream16CreateGrpcStreamEPNS1_14GrpcConnectionERKNS0_4auth5TokenE + 101
	11  macOS_example                       0x000000010030ab9f _ZN8firebase9firestore6remote6Stream26ResumeStartWithCredentialsERKNS0_4util8StatusOrINS0_4auth5TokenEEE + 351
	12  macOS_example                       0x000000010030f49e _ZZZN8firebase9firestore6remote6Stream18RequestCredentialsEvENK3$_0clERKNS0_4util8StatusOrINS0_4auth5TokenEEEENKUlvE_clEv + 142
	13  macOS_example                       0x000000010030f3fd _ZNSt3__128__invoke_void_return_wrapperIvE6__callIJRZZN8firebase9firestore6remote6Stream18RequestCredentialsEvENK3$_0clERKNS4_4util8StatusOrINS4_4auth5TokenEEEEUlvE_EEEvDpOT_ + 45
	14  macOS_example                       0x000000010030f0c9 _ZNSt3__110__function6__funcIZZN8firebase9firestore6remote6Stream18RequestCredentialsEvENK3$_0clERKNS3_4util8StatusOrINS3_4auth5TokenEEEEUlvE_NS_9allocatorISE_EEFvvEEclEv + 41
	15  macOS_example                       0x0000000100129606 _ZNKSt3__18functionIFvvEEclEv + 102
	16  macOS_example                       0x00000001000585ee _ZN8firebase9firestore4util10AsyncQueue15ExecuteBlockingERKNSt3__18functionIFvvEEE + 478
	17  macOS_example                       0x000000010005b177 _ZZN8firebase9firestore4util10AsyncQueue4WrapERKNSt3__18functionIFvvEEEENK3$_0clEv + 39
	18  macOS_example                       0x000000010005b13d _ZNSt3__128__invoke_void_return_wrapperIvE6__callIJRZN8firebase9firestore4util10AsyncQueue4WrapERKNS_8functionIFvvEEEE3$_0EEEvDpOT_ + 45
	19  macOS_example                       0x000000010005aef9 _ZNSt3__110__function6__funcIZN8firebase9firestore4util10AsyncQueue4WrapERKNS_8functionIFvvEEEE3$_0NS_9allocatorISB_EES7_EclEv + 41
	20  macOS_example                       0x0000000100129606 _ZNKSt3__18functionIFvvEEclEv + 102
	21  macOS_example                       0x000000010009b941 _ZZN8firebase9firestore4util8internal13DispatchAsyncEPU28objcproto17OS_dispatch_queue8NSObjectONSt3__18functionIFvvEEEENK3$_0clEPv + 33
	22  macOS_example                       0x000000010009b918 _ZZN8firebase9firestore4util8internal13DispatchAsyncEPU28objcproto17OS_dispatch_queue8NSObjectONSt3__18functionIFvvEEEEN3$_08__invokeEPv + 24
	23  libdispatch.dylib                   0x000000010160c7c3 _dispatch_client_callout + 8
	24  libdispatch.dylib                   0x00000001016149c7 _dispatch_lane_serial_drain + 789
	25  libdispatch.dylib                   0x000000010161587e _dispatch_lane_invoke + 446
	26  libdispatch.dylib                   0x0000000101620f44 _dispatch_workloop_worker_thread + 691
	27  libsystem_pthread.dylib             0x0000000101687047 _pthread_wqthread + 409
	28  libsystem_pthread.dylib             0x0000000101686e41 start_wqthread + 13
```

#2241 added code that should have addressed this but it's not working in a clean install.

### PR DESCRIPTION
**Pull request by wilhuff**
The prior implementation happened to work but only because it made implicit assumptions about the structure of the bundles we load.

In particular, by looking for `@"gRPCCertificates.bundle/roots"` the code was assuming the iOS bundle layout for the app was in effect, but this doesn't work on macOS, in which app/framework bundles have an additional `Contents/Resources` in the path.

This change gets rid of the implicit mapping of the bundle structure onto the directory structure and instead opens the nested `gRPCCertificates.bundle` itself as a resource, making this code portable without having to understand the bundle layout under the covers.

Fixes #2604. Should also remove the need for the workaround in #2177.

### CODE REVIEW

File: Firestore/Example/App/iOS/Images.xcassets/AppIcon.appiconset/Contents.json [hunk 1]
```diff
@@ -84,10 +84,15 @@
       "idiom" : "ipad",
       "size" : "83.5x83.5",
       "scale" : "2x"
+    },
+    {
+      "idiom" : "ios-marketing",
```
Comment (var-const) [inline Firestore/Example/App/iOS/Images.xcassets/AppIcon.appiconset/Contents.json]: Nit: are changes to this file intentional?

File: Firestore/Example/App/macOS_example/AppDelegate.m [hunk 2]
```diff
@@ -35,7 +35,6 @@ - (void)applicationDidFinishLaunching:(NSNotification *)aNotification {
 
   // do the timestamp fix
   FIRFirestoreSettings *settings = db.settings;
-  settings.timestampsInSnapshotsEnabled = true;
   db.settings = settings;
```
Comment (var-const) [inline Firestore/Example/App/macOS_example/AppDelegate.m]: This change seems unrelated?

File: Firestore/Example/Firestore.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -561,7 +561,7 @@
 		B9C261C26C5D311E1E3C0CB9 /* query_test.cc */ = {isa = PBXFileReference; includeInIndex = 1; lastKnownFileType = sourcecode.cpp.cpp; path = query_test.cc; sourceTree = "<group>"; };
 		BB92EB03E3F92485023F64ED /* Pods_Firestore_Example_iOS_Firestore_SwiftTests_iOS.framework */ = {isa = PBXFileReference; explicitFileType = wrapper.framework; includeInIndex = 0; path = Pods_Firestore_Example_iOS_Firestore_SwiftTests_iOS.framework; sourceTree = BUILT_PRODUCTS_DIR; };
```
Comment (var-const) [inline Firestore/Example/Firestore.xcodeproj/project.pbxproj]: Nit: are changes to this file intended as part of this PR?

File: Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder_apple.mm [hunk 4]
```diff
@@ -16,14 +16,17 @@
 
 #include "Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder.h"
 
+#import <objc/runtime.h>
+
 #include <string>
 
+#import "Firestore/Source/Core/FSTFirestoreClient.h"
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder_apple.mm]: Optional: I'm not sure this `import` is necessary, maybe try just deleting it?

File: Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder_apple.mm [hunk 5]
```diff
@@ -34,45 +37,107 @@
...
+  Class firestore_class = objc_getClass("FIRFirestore");
+  if (!firestore_class) return nil;
+
+  return [NSBundle bundleForClass:firestore_class];
+}
+
+/**
+ * Finds the path to the roots.pem certificates file, wherever it may be.
+ *
+ *
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder_apple.mm]: Ultranit: extra empty line.

File: Firestore/Example/App/iOS/Images.xcassets/AppIcon.appiconset/Contents.json (hunk 1 above)
Comment (wilhuff) [reply Firestore/Example/App/iOS/Images.xcassets/AppIcon.appiconset/Contents.json]: Yes. They happen any time you manipulate the macOS project. I can put this in a separate PR if you prefer.

File: Firestore/Example/App/macOS_example/AppDelegate.m (hunk 2 above)
Comment (wilhuff) [reply Firestore/Example/App/macOS_example/AppDelegate.m]: This shows up as a deprecation warning while building the macOS_example, which was the primary means by which I verified this change was correct.

I can move this into a separate cleanup grab-bag PR if you prefer.

File: Firestore/Example/Firestore.xcodeproj/project.pbxproj (hunk 3 above)
Comment (wilhuff) [reply Firestore/Example/Firestore.xcodeproj/project.pbxproj]: As above, these are changes that are happening whenever you touch the project.

File: Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder_apple.mm (hunk 4 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder_apple.mm]: Done. I think we had this before in order to find a class to load the FirebaseFirestore framework's bundle, but doing this reflectively seems cleaner from a layering perspective.

File: Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder_apple.mm (hunk 5 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder_apple.mm]: Ultradone.

**Comment by wilhuff** [comment]
Several of the things you pointed out were unrelated to the core change here though I've included them for convenience. If you prefer I can split these into a separate PR.

File: Firestore/Example/App/macOS_example/AppDelegate.m (hunk 2 above)
Comment (var-const) [reply Firestore/Example/App/macOS_example/AppDelegate.m]: No, I think that would be overkill (applies to other changes of that kind in this PR). As long as this is done on purpose, I'm fine with it being the part of the same PR.

**Comment by var-const** [comment]
I'm okay with these changes being in this PR, was just making sure they were intentional.
